nixos/modules/services/hardware/udev.nix

```diff
@@ -16,16 +16,6 @@ let
   '';
 
 
-  # networkd link files are used early by udev to set up interfaces early.
-  # This must be done in stage 1 to avoid race conditions between udev and
-  # network daemons.
-  # TODO move this into the initrd-network module when it exists
-  initrdLinkUnits = pkgs.runCommand "initrd-link-units" {} ''
-    mkdir -p $out
-    ln -s ${udev}/lib/systemd/network/*.link $out/
-    ${lib.concatMapStringsSep "\n" (file: "ln -s ${file} $out/") (lib.mapAttrsToList (n: v: "${v.unit}/${n}") (lib.filterAttrs (n: _: hasSuffix ".link" n) config.systemd.network.units))}
-  '';
-
```

Comment:
Do these `config.systemd.network.units` link units still get copied to systemd-stage-1 anywhere?

I do not see some similar snippet added anywhere and generating an initrd with `boot.initrd.systemd.enable = true` and `systemd.network.links` (with `boot.initrd.network.enable` both on and off) does not seem to end up with any link units in `/etc/systemd/network` in initrd

Reply:
@lilyinstarlight Good catch. I thought the `systemd.contents = mkUnitFiles "/etc/" cfg;` line was all we needed but that's clearly not quite the same thing since it isn't including all stage 2 link units.

Reply:
@lilyinstarlight I *think* it's ok to just add this to `boot.initrd.systemd.network.units`, [like this](https://github.com/NixOS/nixpkgs/compare/0ff15452acae8164697dcb5bd7247d878ee59dad..3cb9534df64937f3425dbc0f63749cc83276791a).

Reply:
Can we extend / add one test to make sure the logic with an interface rename via .link is exercised?

Maybe also a .netdev file could be used to create a "dummy" interface, and ensure it's kept (probably easier to test than a vlan)

Reply:
I think it's a conceptual question on how much of the network config in initrd should match the one on the real filesystem.

Personally, I'd be in favor of copying all of it in - it's small text files, most of the time the config won't differ much from the real OS, and the user can always put a lexicographically smaller unit file before to override it specifically for in-initrd.

That way, enabling networkd in initrd will give you the same config, which is the one of the least surprises.

At least for the .link files we /need/ to do it though, as doing it after udev started and the interfaces are already brought up will be too late.

Reply:
> Personally, I'd be in favor of copying all of it in

We definitely can't do that, particularly because of things like wireguard that need secret files and whatnot. We could copy in everything that doesn't need support files from the host FS, but I don't know if we have a way to categorically know which those are. Plus there are kernel modules to consider; it's not always clear which kernel modules would need to be added to initrd to support the entirety of someone's networkd config.

That said, you can (and I do) just do `boot.initrd.systemd.network = config.systemd.network` to make them the same.

Anyway, as for the link units, the PR that added this comment also [linked to an issue](https://github.com/systemd/systemd/issues/7293), but that issue was closed by [this](https://github.com/systemd/systemd/pull/11881) way *before* that PR was even made. I'm wondering why that systemd change isn't enough to resolve the problem. *Surely* you're not expected to do your interface renaming in stage 1. In fact, the comments in the issue suggest that the udev rules to rename the interface in stage 2 should happen before `networkd` is even notified the interface is ready.

Reply:
> @lilyinstarlight I _think_ it's ok to just add this to `boot.initrd.systemd.network.units`, [like this](https://github.com/NixOS/nixpkgs/compare/0ff15452acae8164697dcb5bd7247d878ee59dad..3cb9534df64937f3425dbc0f63749cc83276791a). Does that seem good to you?

Yep! That diff looks good to me

> But again, that's probably a concern for a future PR.

I agree there, since it would otherwise be a behavior change and I'd love to get this in within the next day or two as-is